README.md

```diff
@@ -1,16 +1,695 @@
+
+The `out` field MAY omit any tasks that have not yet completed, or results which are not public. In this case, it is RECOMMENDED that 
+
+### 4.1.3 Metadata Fields
+
+The metadata field MAY be omitted or used to contain additional data about the receipt. This field MAY be used for tags, commentary, trace information, and so on.
+
+## 4.2 IPLD Schema
+
+``` ipldsch
+type SignedReceipt struct {
+  rec Receipt (rename "ucan/receipt")
+  sig VarSig
+}
+
+type Receipt struct {
+  inv  &SignedInvocation
+  out  {String : Result}
+  meta Any (implicit Null)
+} 
+
+type Result union {
+  | Payload Any
+  | &SignedReceipt
+}
+```
+
+## 4.3 JSON Example
+
+``` json
+{
+  "ucan/receipt": {
+    "v": "0.1.0",
+    "inv": [{"/": "bafkreifzjut3te2nhyekklss27nh3k72ysco7y32koao5eei66wof36n5e"}],
+    "out": {
+      "bafkreiakkqwuffzsxrzseo7fweeicqlnqanhvyffjieh3etsbnnkjxbphi": [
+        {
+          "from": "alice@example.com",
+          "text": "Hello world!"
+        }, 
+        {
+          "from": "bob@example.com",
+          "text": "What's up?"
+        }
+      ],
+      "bafkreienxymjwglxlb3rkdyeyjt54ddoe4x4qi7a7hyfce3z56zspmy6pm": {
+        "http": { 
+          "status": 200,
+          "body": "hello world"
+        },
+        "ms": 476
+      },
+      "delegated-task": {"/": "bafkreieiupg4smeb5ammpsydbiea4yvwzwne5ly4hiripy4cjocqiat3ce"}
+    }
+    "meta": {
+      "notes": "very receipt. such wow.",
+      "tags": ["dag-house", "fission"]
+    }
+  },
+  "sig": {"/": {"bytes": "bdNVZn_uTrQ8bgq5LocO2y3gqIyuEtvYWRUH9YT-SRK6v_SX8bjt_VZ9JIPVTdxkWb6nhVKBt6JGpgnjABpOCA"}}
+ }
+```
+
+# 5 Promise Pipelining
+
+> Machines grow faster and memories grow larger. But the speed of light is constant and New York is not getting any closer to Tokyo. As hardware continues to improve, the latency barrier between distant machines will increasingly dominate the performance of distributed computation. When distributed computational steps require unnecessary round trips, compositions of these steps can cause unnecessary cascading sequences of round trips
+>
+> — [Mark Miller](https://github.com/erights), [Robust Composition](http://www.erights.org/talks/thesis/markm-thesis.pdf)
+
+At UCAN creation time, the UCAN MAY not yet have all of the information required to construct the next request in a sequence. Waiting for each request to complete before proceeding to the next task has a performance impact due to the amount of latency. [Promise pipelining](http://erights.org/elib/distrib/pipeline.html) is a solution to this problem: by referencing a prior invocation, a pipelined invocation can direct the executor to use the output of earlier invocations into the input of a later one. This liberates the invoker from waiting for each step.
```

Comment:
I feel like this is mixing up invocations and tasks. I think it would really help to clearly separate individual tasks (which I used to call invocations) from the batch (which we call execution). I think here you mean that promises can refer to individual task results as opposed to result of the batch.

Reply:
Fair point! I also have asked myself this question, and it even goes back to some of the diagrams in the presentations:

<img width="542" alt="Screenshot 2022-12-02 at 13 35 39" src="https://user-images.githubusercontent.com/1052016/205392927-85b12fea-7389-4a57-bd07-4601b7882265.png">

Will fix 👍

Reply:
For what it’s worth we call thing you call task an invocation and what you call invocation a query in our implementation. Our current implementation represents query as a tuple of invocations & intention was that we would add more advanced syntax like graph queries in a future iteration. In a way tuple is just a very trivial query where results are mapped by index